File: leetcode beginners/subtree of another tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSameTree(self, b, q):
        if b == None or q == None:
            return b == q

        return self.isSameTree(b.left, q.left) and self.isSameTree(b.right, q.right) and b.val == q.val


    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        if root == None:
            return False

        if root.val == subRoot.val and self.isSameTree(root, subRoot):
            return True

        return self.isSubtree(root.left, subRoot) or self.isSubtree(root.right, subRoot)

File: leetcode in c/subtree of another tree.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
bool isSameTree(struct TreeNode* p, struct TreeNode* q)
{
    if (p == NULL || q == NULL)
    {
        return p == q;
    }

    bool isLeft = isSameTree(p->left, q->left);
    bool isRight = isSameTree(p->right, q->right);

    return isLeft && isRight && p->val == q->val;
}

bool isSubtree(struct TreeNode* root, struct TreeNode* subRoot) {
    if (root == NULL)
    {
        return false;
    }

    if (root->val == subRoot->val && isSameTree(root, subRoot))
    {
        return true;
    }


    return isSubtree(root->left, subRoot) || isSubtree(root->right, subRoot);

}